src/index.js

```diff
@@ -183,20 +183,56 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function visitTailExpressions(path, allowLoops, visitor) {
+    path.resync();
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+    let tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (let tailPath of tailPaths) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```

Comment:
Try these in your console: 

```
function outer(x) {
  x = inner();
  function inner() {
    return x + 1
  }
  return inner
}
outer(1)()
// 3
```
function outer2(x) {
  x = inner();
  return function inner() {
    return x + 1
  }
}
outer2(1)()
// Uncaught ReferenceError: inner is not defined
```